hibernate-core/src/main/java/org/hibernate/dialect/FirebirdDialect.java

```diff
@@ -198,18 +237,143 @@ public void initializeFunctionRegistry(QueryEngine queryEngine) {
+				.setExactArgumentCount( 0 )
+				.setInvariantType( StandardBasicTypes.UUID_BINARY );
+		functionRegistry.namedDescriptorBuilder( "uuid_to_char" )
+				.setExactArgumentCount( 1 )
+				.setInvariantType( StandardBasicTypes.UUID_CHAR );
+		functionRegistry.namedDescriptorBuilder( "gen_id" )
+				.setExactArgumentCount( 2 )
+				.setInvariantType( StandardBasicTypes.BIG_INTEGER )
+				.setArgumentListSignature( "(sequence, step)" );
+
+		if ( getVersion() >= 300 ) {
+			CommonFunctionFactory.stddevPopSamp( queryEngine );
+			CommonFunctionFactory.varPopSamp( queryEngine );
+			functionRegistry.namedDescriptorBuilder( "covar_pop" )
+					.setInvariantType( StandardBasicTypes.DOUBLE )
+					.setExactArgumentCount( 2 )
+					.register();
+			functionRegistry.namedDescriptorBuilder( "covar_samp" )
+					.setInvariantType( StandardBasicTypes.DOUBLE )
+					.setExactArgumentCount( 2 )
+					.register();
+			functionRegistry.namedDescriptorBuilder( "corr" )
+					.setInvariantType( StandardBasicTypes.DOUBLE )
+					.setExactArgumentCount( 2 )
+					.register();
+			Arrays.asList(
+					"regr_avgx", "regr_avgy", "regr_count", "regr_intercept", "regr_r2",
+					"regr_slope", "regr_sxx", "regr_sxy", "regr_syy"
+			)
+					.forEach( fnName ->
+									functionRegistry.namedDescriptorBuilder( fnName )
+											.setInvariantType( StandardBasicTypes.DOUBLE )
+											.setExactArgumentCount( 2 )
+											.register()
+					);
+		}
+
+		if ( getVersion() >= 400 ) {
+			functionRegistry.namedDescriptorBuilder( "hex_encode" )
+					.setExactArgumentCount( 1 )
+					.setInvariantType( StandardBasicTypes.STRING );
+			functionRegistry.namedDescriptorBuilder( "hex_decode" )
+					.setExactArgumentCount( 1 )
+					.setInvariantType( StandardBasicTypes.BINARY );
+			functionRegistry.namedDescriptorBuilder( "base64_encode" )
+					.setExactArgumentCount( 1 )
+					.setInvariantType( StandardBasicTypes.STRING );
+			functionRegistry.namedDescriptorBuilder( "base64_decode" )
+					.setExactArgumentCount( 1 )
+					.setInvariantType( StandardBasicTypes.BINARY );
+
+			functionRegistry.namedDescriptorBuilder( "rdb$system_privilege" )
+					.setExactArgumentCount( 1 )
+					.setInvariantType( StandardBasicTypes.BOOLEAN );
+			functionRegistry.namedDescriptorBuilder( "rdb$role_in_use" )
+					.setExactArgumentCount( 1 )
+					.setInvariantType( StandardBasicTypes.BOOLEAN );
+			functionRegistry.namedDescriptorBuilder( "rdb$get_transaction_cn" )
+					.setExactArgumentCount( 1 )
+					.setInvariantType( StandardBasicTypes.BIG_INTEGER );
+
+			Arrays.asList( "md5", "sha1", "sha256", "sha512" )
+					.forEach( hash -> functionRegistry.registerPattern(
+							hash,
+							"crypt_hash(?1 using " + hash + ")",
+							StandardBasicTypes.BINARY
+					) );
+			functionRegistry.registerAlternateKey( "sha", "sha1" );
+			functionRegistry.registerPattern(
+					"crc32",
+					"hash(?1 using crc32)",
+					StandardBasicTypes.INTEGER
+			);
```

Comment:
I'm don't think I approve of just adding all these. In the other `Dialects I've deliberately got rid of long lists of native functions, and only registered the ones that seemed directly useful in HQL.

(For example, I can't see why I would ever call `rdb$set_context()` from HQL.)

Remember, you can always call any arbitrary function via the JPAQL `function()` syntax.

But it's more than that: it's essentially impossible for us to really maintain these long lists of functions across all the dialects, and we have historically failed to do so in any sort of consistent way, so I don't think we should try.

Reply:
I had somehow assumed that Hibernate would apply the automatic quoting on function names as well if they were not explicitly defined, and then built-in functions will not work (Firebird only supports quoting user-defined function names), but I did some testing and that turns out is not the case. I committed a change where I removed the Firebird specific functions. 

I left the definition of the various SQL standard statistical functions in place. I had considered defining those in `CommonFunctions` like the other statistical functions (e.g. `CommonFunctions.stddevPopSamp(...)`, but I felt that I would then also need to add them to the dialects that also support those functions. 

I also left the mappings for crc32, sha/sha1, md5, sha256 and sha512 in place, as some other dialects define those functions as well. Let me know if you prefer them removed as well.

Reply:
Well md5 and sha functions are defined in `CommonFunctionFactory` so you should reuse those definitions.

I’m fine with adding any SQL-standard-defined statistical functions, but they should also be defined in  `CommonFunctionFactory`.

Reply:
> Well md5 and sha functions are defined in CommonFunctionFactory so you should reuse those definitions.

Oh I see, it’s because you’re mapping them to `crypt_hash`. That’s fine, I suppose.

Reply:
I moved the definition statistical functions into `CommonFunctionFactory` in upcoming commit.

Reply:
And I also added (some of) the statistical functions to Oracle, DB2 and PostgreSQL, in as far as I could quickly find references to them in their online documentation.

Reply:
OK, thanks, that's great.